Add Manuscript Timeline plugin

EricRhysTaylor opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline

## Release Checklist
- [ ] I have tested the plugin on
  - [ ]  Windows
  - [ x]  macOS
  - [ ]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [n/a] `styles.css` _(optional)_
  - [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x ] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x ] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x ] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x ] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x ] I have added a license in the LICENSE file.
- [x ] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

EricRhysTaylor commented:
files added to repository.

EricRhysTaylor commented:
Published 1.0.3 release.

joethei commented:
["css": "styles.css"](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/manifest.json#L10)
This is not a valid part of the manifest, remove it.

[@import "https://fonts.googleapis.com/css2?family=Lato:wght@400;900&display=swap";](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/styles.css#L1)
This should be bundled, not loaded externally, see this guide how to properly embed the fonts
https://docs.obsidian.md/Themes/App+themes/Embed+fonts+and+images+in+your+theme

[.button-container {](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/styles.css#L663-L667), [/* Button container */](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/styles.css#L839-L844)
Please don't overwrite core styling in a plugin, add another class/data attribute so it only applies to your plugin.

[const TIMELINE_VIEW_DISPLAY_TEXT = "Manuscript Timeline";](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L18), ['Manuscript Timeline'](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L1415), [name: 'Open Timeline View',](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L1422), [name: 'Search Timeline',](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L1428), [name: 'Clear Timeline Search',](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L1434), [.setName('Source Path')](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4763), [.setName('Target Completion Date')](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4774), [.setName('Debug Mode')](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4812), [containerEl.createEl('h2', {text: 'Publishing Stage Colors', cls: 'setting-item-heading'});](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4823)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[// Add horizontal rule to separate settings](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4754-L4758)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

[containerEl.createEl('h2', {text: 'Publishing Stage Colors', cls: 'setting-item-heading'});](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4823), 

[await MarkdownRenderer.renderMarkdown(](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4892-L4897), [await MarkdownRenderer.renderMarkdown(](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4892-L4897)
This function is deprecated, it will also throw an error in the dev console every time it is used., use `MarkdownRenderer.render` instead.

 const localContent = await adapter.read(localReadmePath);
This file will never exist in a prod environment, Obsidian will not download the README into the plugin folder.

[const githubReadmeUrl = 'https://raw.githubusercontent.com/ericrhystaylor/Obsidian-Manuscript-Timeline/refs/heads/master/README.md';](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/eda9ee19544ed26fbc092e197d064260b91f8cb2/main.ts#L4885)
Please bundle this with the plugin and don't make unecessary network calls.

EricRhysTaylor commented:
Hi Johannes,

Thank you for your assistance. I have addressed all of the issues listed. Please let me know if there are any other issues for me to address. Have a nice weekend. 

Eric

joethei commented:
[ this.plugin ](https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/356654a8b357c23b7a05dae9bef9881bc0183758/main.ts#L5360)
The component passed to `MarkdownRenderer.render` should not be the plugin instance.
Doing that leads to a memory leak, because new event listeners are registered, but never removed.
The component is responsible for the lifecycle handling of the event listeners related to the rendered result, so it should be unloaded once there is nothing being rendered anymore.

EricRhysTaylor commented:
Hi Johannes,

Thank you for catching this. I've updated the code. Let me know if there is
anything else. Have a great weekend. -eric

this.plugin was replaced with

this as unknown as Component
<https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/4238cc8c3c045b9776668968c6ee86cb0998f705/main.ts#L5405>

This change should resolve the memory leak issue.

Details:

1.  **The Core Change:** The essential fix was changing the last argument
of `MarkdownRenderer.render` from `this.plugin` to `this`. This ensures
that the lifecycle of the rendered Markdown content (and any associated
event listeners created by the renderer) is tied to the
`ManuscriptTimelineSettingTab` instance itself.
2.  **Lifecycle Management:** Since `ManuscriptTimelineSettingTab` extends
`PluginSettingTab`, which extends `Component`, Obsidian manages its
lifecycle. When the user closes the settings tab, Obsidian will unload this
specific component.
3.  **Cleanup:** By passing `this` as the component context,
`MarkdownRenderer.render` registers its cleanup routines (like removing
event listeners) to be executed when *this specific settings tab component*
is unloaded.
4.  **The Cast (`as unknown as Component`):** This part was purely to
satisfy the TypeScript compiler's type checking. It doesn't change the
underlying JavaScript behavior at runtime. We're still passing the settings
tab instance, which is what matters for fixing the leak.

So, when the settings tab is closed and its component is unloaded, the
resources used by the `MarkdownRenderer` for that specific instance should
now be properly released, preventing the memory leak.

On Thu, Apr 24, 2025 at 12:11 PM Johannes Theiner ***@***.***>
wrote:

> *joethei* left a comment (obsidianmd/obsidian-releases#5782)
> <https://github.com/obsidianmd/obsidian-releases/pull/5782#issuecomment-2828618624>
>
> this.plugin
> <https://github.com/EricRhysTaylor/Obsidian-Manuscript-Timeline/blob/356654a8b357c23b7a05dae9bef9881bc0183758/main.ts#L5360>
> The component passed to MarkdownRenderer.render should not be the plugin
> instance.
> Doing that leads to a memory leak, because new event listeners are
> registered, but never removed.
> The component is responsible for the lifecycle handling of the event
> listeners related to the rendered result, so it should be unloaded once
> there is nothing being rendered anymore.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/obsidianmd/obsidian-releases/pull/5782#issuecomment-2828618624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BEFCPC6D2HP463YIRAGRIWD23EZPNAVCNFSM6AAAAABZRGL44CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRYGYYTQNRSGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

joethei commented:
The settings tab is not a component, and casting it won't make it one either.
It does not have a `onunload` function, which needs to be called for the cleanup to work properly.
Just create a new component when the settings tab is rendered, and unload that component when it is close.d

I also see a bunch of strings that are still not using sentence case.
